File: agent-python/agent/core/sse.py
"""SSE (Server-Sent Events) 格式化工具

标准化协议：所有事件统一绑定 requestId / userId / conversationId / messageId，
确保 Java 后端能正确归并 chunk 到对应的 assistant message。

事件类型：
- start: 流式开始，Java 收到后立即建立消息映射
- chunk: AI 内容片段（含 reasoning 可选）
- done: 对话完成
- error: 错误信息
- ping: 心跳（定时发送）
"""
import json
import time
from typing import Any


def format_sse(
    data: dict[str, Any],
    event: str = "message",
    event_id: str | None = None,
) -> str:
    """
    将数据格式化为 SSE 协议格式

    Args:
        data: 要发送的数据字典
        event: 事件类型（start/chunk/done/error/ping）
        event_id: 事件 ID（可选，格式如 "chunk-0", "done-2"）

    Returns:
        SSE 格式的字符串
    """
    lines = []

    if event_id is not None:
        lines.append(f"id: {event_id}")

    lines.append(f"event: {event}")
    lines.append(f"data: {json.dumps(data, ensure_ascii=False)}")
    lines.append("")  # 空行
    lines.append("")  # 双换行表示消息结束

    return "\n".join(lines)


def _build_base_data(
    event_type: str,
    request_id: str | None = None,
    user_id: int | None = None,
    conversation_id: int | None = None,
    message_id: str | None = None,
) -> dict[str, Any]:
    """
    构建事件的基础数据，统一绑定上下文追踪字段

    所有 SSE 事件（ping 除外）都应调用此方法，
    确保 Java 后端能将 chunk 正确归并到 assistant message。
    """
    data: dict[str, Any] = {"type": event_type}
    if request_id is not None:
        data["requestId"] = request_id
    if user_id is not None:
        data["userId"] = user_id
    if conversation_id is not None:
        data["conversationId"] = conversation_id
    if message_id is not None:
        data["messageId"] = message_id
    return data


def create_start_event(
    request_id: str | None = None,
    user_id: int | None = None,
    conversation_id: int | None = None,
    message_id: str | None = None,
) -> str:
    """
    创建 start 事件（流式开始）

    Java 收到后应立即建立消息映射，准备接收后续 chunk。
    """
    data = _build_base_data("start", request_id, user_id, conversation_id, message_id)
    data["timestamp"] = int(time.time() * 1000)
    return format_sse(data=data, event="start", event_id="start")


def create_chunk_event(
    content: str,
    index: int,
    request_id: str | None = None,
    user_id: int | None = None,
    conversation_id: int | None = None,
    message_id: str | None = None,
    reasoning: str | None = None,
) -> str:
    """
    创建 chunk 事件（AI 内容片段）

    每个 chunk 都绑定完整的上下文，Java 可按 messageId 归并。
    content 始终保证为 str 类型，即使为空字符串也不省略。
    """
    data = _build_base_data("chunk", request_id, user_id, conversation_id, message_id)
    # 防御性处理：确保 content 是 str 类型
    data["content"] = str(content) if content is not None else ""
    data["index"] = index
    if reasoning:
        data["reasoning"] = reasoning
    return format_sse(data=data, event="chunk", event_id=f"chunk-{index}")


def create_done_event(
    request_id: str | None = None,
    user_id: int | None = None,
    conversation_id: int | None = None,
    message_id: str | None = None,
    info: str = "对话完成",
    content_length: int = 0,
    chunk_count: int = 0,
) -> str:
    """
    创建 done 事件（对话完成）

    携带 content_length / chunk_count 供 Java 校验完整性。
    """
    data = _build_base_data("done", request_id, user_id, conversation_id, message_id)
    data["info"] = info
    data["contentLength"] = content_length
    data["chunkCount"] = chunk_count
    data["timestamp"] = int(time.time() * 1000)
    return format_sse(data=data, event="done", event_id="done")


def create_error_event(
    message: str,
    request_id: str | None = None,
    user_id: int | None = None,
    conversation_id: int | None = None,
    message_id: str | None = None,
    index: int | None = None,
    error_code: str | None = None,
) -> str:
    """
    创建 error 事件（错误）

    异常分支也必须绑定完整的上下文，确保 Java 能识别并结束消息。
    """
    data = _build_base_data("error", request_id, user_id, conversation_id, message_id)
    data["message"] = message
    if index is not None:
        data["index"] = index
    if error_code:
        data["errorCode"] = error_code
    data["timestamp"] = int(time.time() * 1000)
    return format_sse(data=data, event="error", event_id="error")


def create_ping_event() -> str:
    """
    创建 ping 事件（心跳）

    用于保持连接活跃，前端可忽略或用于检测连接状态。
    ping 事件不包含上下文（仅用于连接保活）。
    """
    data = {"type": "ping"}
    return format_sse(data=data, event="ping")

File: agent-frontend/src/services/sse.ts
import {
  sseEventDataSchema,
  streamChatParamsSchema,
  type SSEChunkData,
  type SEDoneData,
  type SEEErrorData,
  type SEPingData,
  type SSEEventData,
  type SSEMessageIdData,
} from '../schemas';
import { getAuthStatus, getToken, redirectToLogin } from './auth';

export type {
  SSEChunkData,
  SEDoneData,
  SEEErrorData,
  SEPingData,
  SSEMessageIdData,
};

export type SSEHandlers = {
  onChunk?: (content: string, index: number) => void;
  onReasoning?: (reasoning: string, index: number) => void;
  onDone?: (data: SEDoneData) => void;
  onError?: (error: SEEErrorData) => void;
  onConversationId?: (conversationId: number) => void;
  onMessageId?: (messageId: string) => void;
};

export function parseSSEBlock(block: string): SSEEventData | null {
  const trimmed = block.trim();
  if (!trimmed) {
    return null;
  }

  let eventType = 'chunk';
  let dataStr = '';

  const lines = trimmed.split('\n');
  for (const line of lines) {
    const currentLine = line.trim();

    if (currentLine.startsWith('event: ')) {
      eventType = currentLine.slice(7).trim();
    } else if (currentLine.startsWith('data: ')) {
      dataStr = currentLine.slice(6);
    } else if (currentLine.startsWith('data:')) {
      dataStr = currentLine.slice(5);
    }
  }

  if (!dataStr && trimmed.startsWith('data: ')) {
    dataStr = trimmed.slice(6);
  }

  if (!dataStr) {
    return null;
  }

  try {
    const parsed = JSON.parse(dataStr) as unknown;
    const candidate =
      parsed && typeof parsed === 'object'
        ? {
            ...(parsed as Record<string, unknown>),
            type:
              'type' in (parsed as Record<string, unknown>)
                ? (parsed as Record<string, unknown>).type
                : eventType,
          }
        : { type: eventType };

    const result = sseEventDataSchema.safeParse(candidate);
    if (!result.success) {
      console.warn('[SSE] Failed to validate SSE block:', trimmed, result.error.flatten());
      return null;
    }

    return result.data;
  } catch (error) {
    console.warn('[SSE] Failed to parse SSE block:', trimmed, error);
    return null;
  }
}

export async function readSSEStream(
  response: Response,
  handlers: SSEHandlers,
  signal: AbortSignal,
  options?: {
    heartbeatTimeout?: number;
  }
): Promise<void> {
  if (!response.body) {
    handlers.onError?.({ type: 'error', message: '服务器未返回流式数据' });
    return;
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = '';
  const heartbeatTimeout = options?.heartbeatTimeout ?? 120_000;
  let lastDataTime = Date.now();
  let settled = false;

  const heartbeatTimer = setInterval(() => {
    if (!settled && Date.now() - lastDataTime > heartbeatTimeout) {
      clearInterval(heartbeatTimer);
      settled = true;
      handlers.onError?.({
        type: 'error',
        message: '连接超时，服务器长时间未返回数据，请重试',
      });
      reader.cancel();
    }
  }, Math.min(heartbeatTimeout, 30_000));

  try {
    while (true) {
      const { done, value } = await withTimeout(
        reader.read(),
        heartbeatTimeout,
        '读取流式数据超时'
      );

      if (done) {
        if (buffer.trim()) {
          const eventData = parseSSEBlock(buffer);
          if (eventData) {
            settled = dispatchSSEEvent(eventData, handlers) || settled;
          }
        }

        if (!settled) {
          settled = true;
          handlers.onDone?.({ type: 'done', conversationId: 0, messageId: '', info: '' });
        }

        console.debug('[SSE] Stream closed by server');
        break;
      }

      lastDataTime = Date.now();

      const chunk = decoder.decode(value, { stream: true });
      buffer += chunk;

      const blocks = buffer.split('\n\n');
      buffer = blocks.pop() || '';

      for (const block of blocks) {
        const eventData = parseSSEBlock(block);
        if (!eventData) {
          continue;
        }

        settled = dispatchSSEEvent(eventData, handlers) || settled;
        if (settled) {
          console.debug('[SSE] Stream settled by event, stop reading loop');
          break;
        }
      }

      if (settled) {
        break;
      }
    }
  } catch (error) {
    clearInterval(heartbeatTimer);

    if (signal.aborted) {
      console.debug('[SSE] Stream aborted');
      return;
    }

    if (error instanceof Error && error.message === '读取流式数据超时') {
      handlers.onError?.({
        type: 'error',
        message: '连接超时，服务器长时间未返回数据，请重试',
      });
      return;
    }

    let errorMessage = '读取流式数据失败';
    if (error instanceof TypeError) {
      errorMessage = '网络连接异常，请检查网络后重试';
    } else if (error instanceof Error) {
      const message = error.message.toLowerCase();
      if (
        message.includes('network') ||
        message.includes('connection') ||
        message.includes('reset')
      ) {
        errorMessage = '网络连接中断，请重试';
      } else {
        errorMessage = error.message;
      }
    }

    handlers.onError?.({ type: 'error', message: errorMessage });
  } finally {
    clearInterval(heartbeatTimer);
  }
}

function withTimeout<T>(
  promise: Promise<T>,
  timeoutMs: number,
  message: string
): Promise<T> {
  return new Promise<T>((resolve, reject) => {
    const timer = setTimeout(() => reject(new Error(message)), timeoutMs);
    promise.then(
      (value) => {
        clearTimeout(timer);
        resolve(value);
      },
      (error) => {
        clearTimeout(timer);
        reject(error);
      }
    );
  });
}

export async function createSSEConnection(
  message: string,
  conversationId: number | undefined,
  handlers: SSEHandlers,
  signal: AbortSignal
): Promise<void> {
  const authStatus = getAuthStatus();
  if (authStatus === 'expired') {
    handlers.onError?.({ type: 'error', message: '登录已过期，请重新登录' });
    redirectToLogin('session-expired');
    return;
  }

  if (authStatus === 'invalid') {
    handlers.onError?.({ type: 'error', message: '登录信息无效，请重新登录' });
    redirectToLogin('unauthorized');
    return;
  }

  const requestBody = streamChatParamsSchema.parse({
    message,
    ...(conversationId !== undefined ? { conversationId } : {}),
  });

  let response: Response;
  try {
    const token = getToken();
    response = await fetch('/agent/chat/stream', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...(token ? { Authorization: `Bearer ${token}` } : {}),
      },
      body: JSON.stringify(requestBody),
      signal,
    });
  } catch (error) {
    if (signal.aborted) {
      console.debug('[SSE] Request aborted');
      return;
    }

    let errorMessage = '连接失败';
    if (error instanceof TypeError && error.message.includes('fetch')) {
      errorMessage = '网络连接失败，请检查网络后重试';
    } else if (error instanceof DOMException && error.name === 'AbortError') {
      console.debug('[SSE] Request aborted by signal');
      return;
    } else if (error instanceof Error) {
      const networkMessage = error.message.toLowerCase();
      if (
        networkMessage.includes('network') ||
        networkMessage.includes('connection') ||
        networkMessage.includes('reset')
      ) {
        errorMessage = '网络连接中断，请重试';
      } else {
        errorMessage = `连接失败: ${error.message}`;
      }
    }

    handlers.onError?.({ type: 'error', message: errorMessage });
    return;
  }

  if (!response.ok) {
    let errorMessage =
      response.status === 401
        ? '认证失败，请重新登录'
        : response.status === 403
          ? '无权访问当前会话'
          : response.status === 429
            ? '请求过于频繁，请稍后重试'
            : `服务器错误 (${response.status})`;

    if (response.status === 401) {
      try {
        const errorData = await response.json();
        errorMessage = errorData?.message || errorMessage;
      } catch {
        // keep fallback message
      }

      handlers.onError?.({ type: 'error', message: errorMessage });
      redirectToLogin(errorMessage.includes('过期') ? 'session-expired' : 'unauthorized');
      return;
    }

    if (response.status === 403) {
      try {
        const errorData = await response.json();
        errorMessage = errorData?.message || errorMessage;
      } catch {
        // keep fallback message
      }
    }

    handlers.onError?.({ type: 'error', message: errorMessage });
    return;
  }

  await readSSEStream(response, handlers, signal, {
    heartbeatTimeout: 120_000,
  });
}

export function dispatchSSEEvent(
  eventData: SSEEventData,
  handlers: SSEHandlers
): boolean {
  switch (eventData.type) {
    case 'message_id':
      if (eventData.messageId) {
        handlers.onMessageId?.(eventData.messageId);
      }
      return false;
    case 'chunk': {
      const reasoningContent = eventData.reasoning || eventData.info;
      if (reasoningContent) {
        handlers.onReasoning?.(reasoningContent, eventData.index);
      }

      if (eventData.content) {
        handlers.onChunk?.(eventData.content, eventData.index);
      }

      if (typeof eventData.conversationId === 'number') {
        handlers.onConversationId?.(eventData.conversationId);
      }

      return false;
    }
    case 'done':
      handlers.onConversationId?.(eventData.conversationId);
      handlers.onDone?.(eventData);
      return true;
    case 'error':
      handlers.onError?.(eventData);
      return true;
    case 'ping':
      console.debug('[SSE] ping received');
      return false;
    default:
      return false;
  }
}
